src/System.Private.CoreLib/shared/System/Number.Formatting.cs

```diff
@@ -2376,7 +2376,13 @@ internal static unsafe void RoundNumber(ref NumberBuffer number, int pos)
             while (i < pos && dig[i] != 0)
                 i++;
 
-            if (i == pos && dig[i] >= '5')
+            // We only want to round up if the digit is greater than or equal to 5 and we are
+            // not rounding a floating-point number. This is because the floating-point format
+            // algorithms already produce a correctly rounded result. They may, however, still
+            // produce trailing zeros that we don't currently display and so we still utilize
+            // the logic in the else block to trim those.
+
+            if ((i == pos) && (dig[i] >= 5) && (number.Kind != NumberBufferKind.FloatingPoint))
```

Comment:
This would be correct, except for the custom format string case, where we default to a precision of 17 and depend on this call to round it correctly.

That is problematic though because it can induce double-rounding (we round to 17 digits and then to the desired digit count). It's also problematic because if the user requested a custom format string with 20 digits, we would only return 17 (or fill in with 0, depending on the format string).

The ideal fix would be to pre-parse the custom format string and get the appropriate digit count, which would let Dragon4/Grisu3 get the correct digit count and round the buffer correctly.

However, the more trivial partial-fix is to check if there is a digit past `pos` and use that to impact the `dig[i]` == 5` case (this still has potential for double-rounding on custom-format strings).